ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java

```diff
@@ -8810,7 +8810,8 @@ private Operator genConversionSelectOperatorByAddPartition(String dest, QB qb, O
     List<ExprNodeDesc> expressions = new ArrayList<>(outColumnCnt);
 
     AtomicBoolean convert = new AtomicBoolean(false);
-    if (inColumnCnt + staticPartitionColumnCnt != outColumnCnt) {
+    if (inColumnCnt + staticPartitionColumnCnt != outColumnCnt &&
```

Comment:
i don't think we should just ignore that check for iceberg, can we update the `staticPartitionColumnCnt`?
cc @kasakrisz

Reply:
I modified the check for Iceberg case as following, is it fine now?

```
    if (table.getStorageHandler() != null && table.getStorageHandler().alwaysUnpartitioned()) {
      if (inColumnCnt != outColumnCnt || staticPartitionColumnCnt > inColumnCnt){
        String reason = "Always unpartitioned table " + dest + " has " + outColumnCnt
            + " columns, but query has " + inColumnCnt + " columns.";
        throw new SemanticException(ASTErrorUtils.getMsg(
            ErrorMsg.TARGET_TABLE_COLUMN_MISMATCH.getMsg(),
            qb.getParseInfo().getDestForClause(dest), reason));
      }
    }
```

Reply:
previously this check was done regardless of `alwaysUnpartitioned` flag and now we moved it inside the if?

Reply:
we shouldn't overload existing code it's already complicated. is it possible to extract this column check in some validateColumns method?
cc @kasakrisz

Reply:
Moved this validation to a separate method.